### DISCUSSION TITLE
Improvements in LSP API

### DISCUSSION THREAD

**Discussion by BladeMF** [phpactor/phpactor#2255] [General]
List leaving this here, it was in the most recent update:

> # Workspace edits can now create files directly from DataTransferFile
> One of the primary uses of the [drop into editor API](https://github.com/microsoft/vscode-extension-samples/tree/main/drop-on-document) is writing dropped files/content into the workspace. However in previous VS Code releases, this could be fairly slow for large files. This is because the file contents end up being copied between processes twice: first from the renderer to the extension host to read the file contents, and then back from the extension host to the renderer to write the file.
> ```
> class CreateFileDropProvider implements vscode.DocumentDropEditProvider {
>     async provideDocumentDropEdits(_document: vscode.TextDocument, _position: vscode.Position, dataTransfer: vscode.DataTransfer, _token: vscode.CancellationToken): Promise<vscode.DocumentDropEdit | undefined> {
>         const pngFile = dataTransfer.get('image/png')?.asFile();
>         if (!pngFile) {
>             return;
>         }
> 
>         // Read file
>         // This results in the entire file contents being copied over to the extension host.
>         const contents = await pngFile.data();
> 
>         // Now create a workspace edit that writes the file into the workspace
>         // This results in the same file contents from above being copied back again.
>         const additionalEdit = new vscode.WorkspaceEdit();
>         const path = vscode.Uri.joinPath(vscode.workspace.workspaceFolders![0].uri, 'image.png');
>         additionalEdit.createFile(path, { contents });
> 
>         const edit = new vscode.DocumentDropEdit(path.fsPath);
>         edit.additionalEdit = additionalEdit;
>         return edit;
>     }
> }
> ```
> Now you can avoid those extra copies though by passing a `DataTransferFile` directly to `WorkspaceEdit.createFile`:
> ```
> additionalEdit.createFile(path, { contents: pngFile });
> ```
> This should significantly improve performance, especially when working with larger files.
> 
> # Resolve Code Action commands in `resolveCodeAction`
> A `CodeActionProvider` can now lazily resolve the command of `CodeAction` in `resolveCodeAction`. Previously only the edits for the Code Action could be lazily resolved.
> 
> If the command is expensive to compute, this allows a `CodeActionProvider` to defer this work until the Code Action is going to be applied.
> 
> `editor/lineNumber/context menu`
> We have finalized the `editor/lineNumber/context menu`. This allows extension authors to contribute actions to a context menu anchored to the editor line number and glyph margin. Actions contributed to this menu receive the line number in command arguments and can reference the editorLineNumber context key in their when clauses.

**Comment by dantleech** [comment]
~hey, what is the source? 3.17 is the most recent version and the one we support: https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/~ i guess this relates to the VSCode API rather than LSP?
